### PR TITLE
Setting up revision previews with Ember CLI deploy

### LINKED ISSUE #450: Ember CLI Deploy: Revision previews
I've set up a project to have `/rev-1abc322` preview the `1abc322` commit. I could write a post about it.

### PR DESCRIPTION
**Pull request by martndemus**
Closes #450

I'm not a native english speaker, spellcheck please.

### CODE REVIEW

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+title: "Setting up revision previews with Ember CLI Deploy"
+author: 'Marten Schilstra'
+twitter: 'martndemus'
+github: martndemus
+published: true
+tags: ember
+---
+
+I recently set up previewing revisions with [Ember CLI Deploy][ember-cli-deploy].
+There is not a single best practice on how to do this yet, but here is how I did
```
Comment (bcardarella) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: "There is not a single" => There is no single"

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+it.
+
+_This post assumes you use the
+[ember-cli-deploy-revision-data][ember-cli-deploy-revision-data] plugin in
+conjunction with an index adapter for Ember CLI Deploy._
+
+### About revisions
+
+Everytime you deploy `index.html` with Ember CLI deploy it gets tagged with a
```
Comment (bcardarella) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: might want to link out to the ember-cli project here

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+### Viewing revisions without having to activate them
+
+It would be very nice to be able preview (maybe QA test) on the production
+server, before you activate it and release it to the public. My idea was to be
+able to do this by visiting `https://example.com/rev-a76df687f97e9ab8ca82d1a1/`.
+
+It is really simple to set up your webserver to do this, so I won't go into
+that, but there is a caveat on the Ember side of this. If you navigate to the
+revision, the router will throw you an error: `The route rev-a76df687f97e9ab8ca82d1a1/
```
Comment (bcardarella) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: "the router will throw you an error" => "the router will throw an error"

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+was not found`.
+
+What is happening? The `Ember.Router` is not intelligent enough to know that you
+meant to route after the `/rev-a76df687f97e9ab8ca82d1a1/` portion of the path.
```
Comment (bcardarella) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Might be beneficial to explain why this is the case if this is known

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+To route revisions is a little more complex, as you can't just type in a simple
+string and be done with it. To be able to handle  revisions, you can make the
+`rootURL` a computed property instead:
+
+```javascript
+const Router = Ember.Router.extend({
+  location: 'auto',
+
+  rootURL: Ember.computed(function() {
```
Comment (bcardarella) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: We prefer to destructure things like `Ember.computed` rather than work from the constant:

```javascript
const {
  computed,
  isNone,
  Router
} = Ember;

export default Router.extend({
  location: 'auto',
  rootURL: computed(() => {
    ...
```

it may seem more verbose (and it is) for just a blog post but we try to push our styleguide when we can. This would include pushing more es2016 features when possible.

**Comment by bcardarella** [comment]
Looks good. I think you're missing the outbound link references and there is that TODO at the end but :+1: so far!

**Comment by martndemus** [comment]
Updated per feedback, but still need to close on something less of a cliffhanger.

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+it.
+
+_This post assumes you use the
+[ember-cli-deploy-revision-data][ember-cli-deploy-revision-data] plugin in
+conjunction with an index adapter ([Redis][redis-index], [S3][s3-index],
+[SSH][ssh-index]) for Ember CLI Deploy._
+
+### About revisions
+
+Everytime you deploy `index.html` with Ember CLI deploy it gets tagged with a
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Add a comma after "deploy".

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+revision. By default this is an MD5 hash of the `index.html` file itself, but it
+can also be a Git commit hash, or the version from your `package.json`.
+
+It depends on the index adapter how the `index.html` get tagged exactly. For
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: I'd reword it like this: "How exactly `index.html` gets tagged depends on the index adapter."

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+example: using the S3 adapter, it will append the tag to the filename; using
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: "Using the S3 adapter, the tag will be appended to the filename" -- just to mirror the Redis statement that follows, and also because currently, "it" in "it will append" is ambiguous.

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+Redis, the tag will be part of the key where the file is stored.
+
+Once you have one or more revisions deployed to the target server, you can
+activate one of the revisions, which will be the default `index.html`
+the users of your app will see.
+
+### Viewing revisions without having to activate them
+
+It would be very nice to be able preview (maybe QA test) on the production
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: "to be able to preview **a revision**"

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+was not found`.
+
+What is happening? The `Ember.Router` is not intelligent enough to know that you
+meant to route after the `/rev-a76df687f97e9ab8ca82d1a1/` portion of the path.
+
+You have to explicitly configure the router to route on a path other than `/`.
+
+### Fixing the router
+
+It's pretty simple to set up the router to work from a base path, for example
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Maybe change up the wording as it's a bit too similar to how you start your sentence on line 40, in "It is really simple to set up your webserver...". Something like this could work: "Setting up the router to work from a base path should be fairly straightforward. For example, [...]".

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+const { Router } = Ember;
+
+const Router = Router.extend({
+  location: 'auto',
+  rootURL: `/awesome-app/`
+});
+```
+
+To route revisions is a little more complex, as you can't just type in a simple
+string and be done with it. To be able to handle  revisions, you can make the
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Remove extra space between "handle" and "revisions".

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+`rootURL` a computed property instead:
+
+```javascript
+import ember from 'ember';
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Capitalize `Ember`.

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  rootURL: computed(() => {
+    let path = window.location.pathname;
+
+    // Looks for /rev-a76df687f97e9ab8ca82d1a1/ at the beginning of the path
+    // Tweak this regex to match your own style of revisions.
+    let revisionMatch = new RegExp('^/(rev-.{20}))').exec(path);
+
+    // If there was a revision at the beginning of the path
+    // return it as rootURL
+    if (revisionMatch && !isNone(revisionMatch[1])) {
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: `isPresent` might be clearer than `!isNone`?

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+const Router = Router.extend({
+  location: 'auto',
+
+  rootURL: computed(() => {
+    let path = window.location.pathname;
+
+    // Looks for /rev-a76df687f97e9ab8ca82d1a1/ at the beginning of the path
+    // Tweak this regex to match your own style of revisions.
+    let revisionMatch = new RegExp('^/(rev-.{20}))').exec(path);
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Even though this is just to illustrate your point, you may want to make the regex actually match the example in your comment. The `a76df687f97e9ab8ca82d1a1` portion is 24 characters, not 20. Also, the regex contains one too many closing parentheses. It should be `new RegExp('^/(rev-.{24})')`.

You could also use a regex literal, e.g.:

```js
let revisionMatch = /^\/(rev-.{24})/.exec(path);
```

I find using the `RegExp` constructor pretty annoying unless you have a dynamic pattern, because you have to double escape everything (not in this specific example, but in general).

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    if (revisionMatch && !isNone(revisionMatch[1])) {
+      return `/${revisionMatch[1]}/`;
+    } else {
+      return '/';
+    }
+  })
+});
+```
+
+When the application loads the `rootURL` computed property will be executed once,
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Comma after "loads".

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md (hunk 14 above)
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: ... or `/^\/(rev-[^/]+)/` if you don't want to count everything.

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md (hunk 15 above)
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Period after "once" instead of a comma. Then start a new sentence: "If the `rev-*` part [...]".

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+if the `rev-` part is in the path it will return that as the `rootURL`, else it
+will just return the default `/`.
+
+### Caveat: baseURL is set in ENV
+
+By default the `baseURL` property in your `config/environment.js` is `/` and
+does not interfere with the `rootURL` in the router, but when it is something
+like `/awesome-app`, then you will run into trouble.
+
+It is best to not use `baseURL` in conjunction with `rootURL`. If you have a bae
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: `s/bae/base/`

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+path the app is served from I would recommend you add it to the `rootURL`.
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Comma after "served".

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+
+Here is how I solved it:
+
+```javascript
+import ember from 'ember';
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: Capitalize `Ember`.

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+const {
+  Router,
+  computed,
+  isNone
+} = Ember;
+
+const Router = Router.extend({
+  location: 'auto',
+
+  rootURL: computed(function() {
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: You're using `function() {}` here but the arrow function in the previous example. Stick with one?

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+    let baseRootURL = config.baseRootURL || ''; // NOTE: This is not baseURL
+    let path = window.location.pathname;
+
+    let revisionMatch = new RegExp(`^${baseRootURL}/(rev-.{20}))`).exec(path);
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: The regex has one too many closing parentheses. Also, change the number to be consistent with the example you're using throughout the post.

**Comment by brzpegasus** [comment]
Good content, @martndemus! This definitely needs a closing paragraph though.

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md (hunk 6 above)
Comment (martndemus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: I'm guessing you meant after "Ember CLI deploy"

**Comment by martndemus** [comment]
Thanks @brzpegasus for the thorough review! I have updated the post per your review, I have also added a closing paragraph.

File: source/posts/2016-02-03-ember-cli-deploy-revision-previews.md [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+const Router = Router.extend({
+  location: 'auto',
+
+  rootURL: computed(() => {
+    let baseRootURL = config.baseRootURL || ''; // NOTE: This is not baseURL
+    let path = window.location.pathname;
+
+    let revisionMatch = new RegExp(`^${baseRootURL}/(rev-[^/])`).exec(path);
+
+    if (revisionMatch && !isNone(revisionMatch[1])) {
```
Comment (brzpegasus) [inline source/posts/2016-02-03-ember-cli-deploy-revision-previews.md]: `isPresent` instead of `!isNone` (since you made the change in the previous example).
